### PR TITLE
M3 add new cache bundle

### PR DESCRIPTION
**Pull request by hluchas**
Commits taken from https://github.com/mautic/mautic/pull/7522 (M2 pull request) and I added latest changes from our cloud version proved by time ;-) 

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | https://github.com/hluchas/mautic/blob/m3-add-new-cache-bundle/app/bundles/CacheBundle/README.md
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This bundle enables the developer to use common Symfony caching mechanism with tag-aware adapters. It implements FS, Memcache and redis adapters.

#### Steps to test this PR:
1. Apply PR + require composer install
2. Set up an adapter as described in documentation
3. Use the cache as described in documentation
4. Verify items are created (e.g. redis records)
5. Verify items expire
6. Clear cache and verify items are purged

### CODE REVIEW

**Comment by hluchas** [comment]
@alanhartless Blame found you as responsible for https://github.com/mautic/mautic/blob/f5455ebc32584039dbad267671d6a5cc4461e257/app/bundles/PageBundle/Controller/PageController.php#L908

which is producing PHPStan error https://travis-ci.org/github/mautic/mautic/jobs/690027759
```
 ------ ------------------------------------------------------ 
  Line   app/bundles/PageBundle/Controller/PageController.php  
 ------ ------------------------------------------------------ 
  896    Undefined variable: $clone                            
 ------ ------------------------------------------------------ 
```

I don't know what is the expected behavior.

This problem is not problem of this PR, but PHPStan reported it and as you can see, there's logical prob ;-)

**Comment by hluchas** [comment]
> @alanhartless Blame found you as responsible for
> 
> https://github.com/mautic/mautic/blob/f5455ebc32584039dbad267671d6a5cc4461e257/app/bundles/PageBundle/Controller/PageController.php#L908
> 
> which is producing PHPStan error https://travis-ci.org/github/mautic/mautic/jobs/690027759
> 
> ```
>  ------ ------------------------------------------------------ 
>   Line   app/bundles/PageBundle/Controller/PageController.php  
>  ------ ------------------------------------------------------ 
>   896    Undefined variable: $clone                            
>  ------ ------------------------------------------------------ 
> ```
> 
> I don't know what is the expected behavior.
> 
> This problem is not problem of this PR, but PHPStan reported it and as you can see, there's logical prob ;-)

TODO: We decided to return 404 when entity is not found.

**Comment by hluchas** [comment]
@escopecz Do you have an idea why this is happening? We hit this prob in the past I think. I don't want to go through fixtures and tests to check where is the difference/what's wrong. This will take some time :-$

```
There was 1 failure:
1) Mautic\LeadBundle\Tests\Segment\ContactSegmentServiceFunctionalTest::testSegmentCountIsCorrect
There should be 53 contacts in the segment-test-5 segment.
Failed asserting that '54' matches expected 53.
/home/travis/build/mautic/mautic/app/bundles/LeadBundle/Tests/Segment/ContactSegmentServiceFunctionalTest.php:56
```

**Comment by escopecz** [comment]
I don't see anything obvious. The PR changes namespace for some campaign fixtures, but the failing test doesn't seem to use them. The only file that can affect it seem to be the CacheHelper.

**Comment by kuzmany** [comment]
@RCheesley I don't see conflicts here. Could we add it to 3.1 a lot of PRs depends on it

**Comment by RCheesley** [comment]
@kuzmany this was being held due to the failing Travis build. If that is addressed then it can be considered.

**Review by kuzmany — DISMISSED** [review]
CacheBundle() should be addedd to AppKernel  asnew bundles https://github.com/mautic/mautic/blob/staging/app/AppKernel.php#L137

After that is work properly. 
I was not able to clear cache by mautic:cache:clear. Maybe it was related to windows test enviroment, because when I was able to clear cache in code `$cache->clear()` 

Then I approve it, bed need add that line to AppKernel.php @hluchas

**Comment by RCheesley** [comment]
@hluchas could you take a look at the comment from @kuzmany above regarding the need to add a line to AppKernel, and the fact that he was unable to clear the cache at command line?

**Comment by hluchas** [comment]
@RCheesley I added cache bundle to kernel. I've probs with local env setup today so hope it works.

**Comment by RCheesley** [comment]
Bumping to 3.2 due to conflicts and Travis build failures.

File: app/bundles/CacheBundle/Config/config.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                'tag'       => 'mautic.cache.adapter',
+            ],
+        ],
+        'models'    => [],
+        'validator' => [],
+    ],
+
+    'parameters' => [
+        'cache_adapter'           => 'mautic.cache.adapter.filesystem',
+        'cache_prefix'            => getenv('DB_NAME') ?: '%mautic.db_name%',
```
Comment (escopecz) [inline app/bundles/CacheBundle/Config/config.php:74]: @hluchas This param seem to be the problem based on the error message from Travis:
```
[RuntimeException]                                                           
  An error occurred when executing the "'cache:clear --no-warmup'" command:    
                                                                               
                                                                               
                                                                               
                                                                               
  In EnvVarProcessor.php line 145:                                             
                                                                               
                                                                               
    Parameter "mautic.db_name" found when resolving env var "MAUTIC_CACHE_PRE  
  FI                                                                           
    X" must be scalar, "NULL" given.  
```

Also, it's weird that the composer.json is adding new libraries but composer.lock has no change

**Comment by hluchas** [comment]
New and complete PR created here https://github.com/mautic/mautic/pull/9107

**Pull request closed without merging**